### DISCUSSION TITLE
Is sessionmaker itself thread-safe?

### DISCUSSION THREAD

**Discussion by andrii-korotkov** [sqlalchemy/sqlalchemy#10232] [Usage Questions] [answered]
Hello. I've read that session objects themselves are not thread-safe, but what about a sessionmaker that would create a new session within each thread. Can it be initialized in the main thread and safely passed to the background threads to create new sessions from? Thank you.

**Comment by zzzeek** [comment, marked as answer]
that's correct, the `sessionmaker()`, once constructed, serves as a factory for new `Session` objects, and has no mutable state beyond when the `.configure()` method is called.  there is no implication to having it called under thread-concurrent scenarios.  typically it's intended to be a module-global object.
